libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -53,18 +53,19 @@ void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
 				" and " +
 				right->toString()
 			);
-		setType(
-			_operation,
+		commonType =
 			TokenTraits::isCompareOp(_operation.getOperator()) ?
 			make_shared<BoolType>() :
-			commonType
-		);
+			commonType;
+		setType(_operation, commonType);
+		_operation.annotation().commonType = commonType;
```

Comment:
Actually the common type is always the one the operation is performed in, so even for comparison operators, it is the common type of its parameters, not bool.

Also, shouldn't the type checker actually set the annotations?

Reply:
I haven't figured out exactly yet which part should do the types, I mean we also set the type in the references resolver: https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/ReferencesResolver.cpp#L249

Reply:
Yes, but only inside typenames.